File: Everyday/p1224.py
from typing import Counter, List


class Solution:
    def maxEqualFreq(self, nums: List[int]) -> int:
        freq, cnt = Counter(), Counter()
        ans = maxFreq = 0
        for i, num in enumerate(nums):
            if cnt[num]:
                freq[cnt[num]] -= 1
            cnt[num] += 1
            freq[cnt[num]] += 1
            maxFreq = max(maxFreq, cnt[num])
            if maxFreq == 1 or\
                    freq[maxFreq] == 1 and maxFreq+freq[maxFreq-1]*(maxFreq-1) == i+1 or\
                    freq[maxFreq]*maxFreq+1 == i+1 and freq[1] == 1:
                ans = max(ans, i+1)
        return ans

File: Everyday/p1224.java
public class p1224 {
    public int maxEqualFreq(int[] nums) {
        int[] cnt = new int[100010];
        int[] freq = new int[100010];
        int ans = 0, maxFreq = 0;
        for (int i = 0; i < nums.length; ++i) {
            int num = nums[i], currCnt = ++cnt[num], len = i + 1;
            ++freq[currCnt];
            --freq[currCnt - 1];
            maxFreq = Math.max(maxFreq, currCnt);
            if (maxFreq == 1)
                ans = len;
            if (freq[maxFreq] == 1 && freq[maxFreq - 1] * (maxFreq - 1) + maxFreq == len)
                ans = len;
            if (freq[1] == 1 && freq[maxFreq] * maxFreq + 1 == len)
                ans = len;
        }
        return ans;
    }
}